src/quota_aggregator_impl.h

```diff
@@ -0,0 +1,180 @@
+                            public AllocateQuotaCacheRemovedItemsHandler {
+ public:
+  // TODO(jaebong) when an element was expired and aggregated tokens are bigger
+  // than 0, aggregator needs to send a request inside the aggregator.
+  QuotaAggregatorImpl(const std::string& service_name,
+                      const std::string& service_config_id,
+                      const QuotaAggregationOptions& options);
+
+  virtual ~QuotaAggregatorImpl();
+
+  // Sets the flush callback function.
+  // The callback function must be light and fast.  If it needs to make
+  // a remote call, it must be non-blocking call.
+  // It should NOT call into this object again from this callback.
+  // It will cause dead-lock.
+  void SetFlushCallback(FlushCallback callback);
+
+  // If the quota could not be handled by the cache, returns NOT_FOUND,
+  // caller has to send the request to service control.
+  // Otherwise, returns OK and cached response.
+  ::google::protobuf::util::Status Quota(
+      const ::google::api::servicecontrol::v1::AllocateQuotaRequest& request,
+      ::google::api::servicecontrol::v1::AllocateQuotaResponse* response);
+
+  // Caches a response from a remote Service Controller AllocateQuota call.
+  ::google::protobuf::util::Status CacheResponse(
+      const ::google::api::servicecontrol::v1::AllocateQuotaRequest& request,
+      const ::google::api::servicecontrol::v1::AllocateQuotaResponse& response);
+
+ private:
+  class CacheElem {
+   public:
+    CacheElem(const ::google::api::servicecontrol::v1::AllocateQuotaResponse&
+                  response)
+        : operation_aggregator_(nullptr), quota_response_(response) {}
+
+    // Aggregates the given request to this cache entry.
+    void Aggregate(
+        const ::google::api::servicecontrol::v1::AllocateQuotaRequest& request);
+
+    // Returns the aggregated AllocateQuotaRequest and reset the cache entry.
+    ::google::api::servicecontrol::v1::AllocateQuotaRequest
+    ReturnAllocateQuotaRequestAndClear(const std::string& service_name,
+                                       const std::string& service_config_id);
+
+    // Change the negative response to the positive response for refreshing
+    void ClearAllocationErrors() { quota_response_.clear_allocate_errors(); }
+
+    bool HasPendingAllocateQuotaRequest() const {
+      return operation_aggregator_ != NULL;
+    }
+
+    // Setter for AllocateQuota response.
+    inline void set_quota_response(
+        const ::google::api::servicecontrol::v1::AllocateQuotaResponse&
+            quota_response) {
+      quota_response_ = quota_response;
+    }
+
+    // Getter for check response.
+    inline const ::google::api::servicecontrol::v1::AllocateQuotaResponse&
+    quota_response() const {
+      return quota_response_;
+    }
+
+    // Return true if aggregated
+    inline bool is_aggregated() const {
+      return operation_aggregator_ ? operation_aggregator_->is_aggregated()
```

Comment:
if each operation has at lease one metric,  we only need to check if operation_aggregator_ is null